action/protocol/staking/validations_test.go

```diff
@@ -155,3 +150,512 @@ func TestProtocol_ValidateCreateStake(t *testing.T) {
+			10000,
+			1,
+			nil,
+		},
+		// an existing owner, but selfstake is not 0
+		{
+			"test2", cans[1].Operator.String(), cans[1].Reward.String(), cans[1].Owner.String(), "100000000000000000000", uint32(10000), false, []byte("payload"), big.NewInt(unit.Qev),
+			10000,
+			1,
+			ErrInvalidOwner,
+		},
+		// IsValidCandidateName special char
+		{"!te", cans[0].Operator.String(), cans[0].Reward.String(), cans[0].Owner.String(), "100000000000000000000", uint32(10000), false, []byte("payload"), big.NewInt(unit.Qev),
+			10000,
+			1,
+			ErrInvalidCanName,
+		},
+		// IsValidCandidateName len>12
+		{"100000000000000000000", cans[0].Operator.String(), cans[0].Reward.String(), cans[0].Owner.String(), "100000000000000000000", uint32(10000), false, []byte("payload"), big.NewInt(unit.Qev),
+			10000,
+			1,
+			ErrInvalidCanName,
+		},
+		// IsValidCandidateName len==0
+		{"", cans[0].Operator.String(), cans[0].Reward.String(), cans[0].Owner.String(), "100000000000000000000", uint32(10000), false, []byte("payload"), big.NewInt(unit.Qev),
+			10000,
+			1,
+			ErrInvalidCanName,
+		},
+		{"test1", cans[0].Operator.String(), cans[0].Reward.String(), cans[0].Owner.String(), "100000000000000000000", uint32(10000), false, []byte("payload"), big.NewInt(-unit.Qev),
+			10000,
+			1,
+			action.ErrGasPrice,
+		},
+		// amount<minSelfStake
+		{
+			"test1", cans[0].Operator.String(), cans[0].Reward.String(), cans[0].Owner.String(), "1", uint32(10000), false, []byte("payload"), big.NewInt(unit.Qev),
+			10000,
+			1,
+			ErrInvalidAmount,
+		},
+		// act.Name() != c.Name && p.inMemCandidates.ContainsName(act.Name())
+		{
+			"test", cans[0].Operator.String(), cans[0].Reward.String(), cans[0].Owner.String(), "100000000000000000000", uint32(10000), false, []byte("payload"), big.NewInt(unit.Qev),
+			10000,
+			1,
+			ErrInvalidCanName,
+		},
+		// !address.Equal(act.OperatorAddress(), c.Operator) && p.inMemCandidates.ContainsOperator(act.OperatorAddress())
+		{
+			"test1", cans[1].Operator.String(), cans[0].Reward.String(), cans[0].Owner.String(), "100000000000000000000", uint32(10000), false, []byte("payload"), big.NewInt(unit.Qev),
+			10000,
+			1,
+			ErrInvalidOperator,
+		},
+		// actCtx.Caller and act.OwnerAddress() is "",p.inMemCandidates.GetByOwner(owner) returns nil,p.inMemCandidates.ContainsName(act.Name())
+		{
+			"test1", cans[0].Operator.String(), cans[0].Reward.String(), "", "100000000000000000000", uint32(10000), false, []byte("payload"), big.NewInt(unit.Qev),
+			10000,
+			1,
+			ErrInvalidCanName,
+		},
+		// actCtx.Caller and act.OwnerAddress() is "",p.inMemCandidates.GetByOwner(owner) returns nil,p.inMemCandidates.ContainsOperator(act.OperatorAddress())
+		{
+			"2222", cans[0].Operator.String(), cans[0].Reward.String(), "", "100000000000000000000", uint32(10000), false, []byte("payload"), big.NewInt(unit.Qev),
+			10000,
+			1,
+			ErrInvalidOperator,
+		},
+	}
```

Comment:
with you latest change, 
we have 4 generic error case (1 is nil action)
3 error case for existing owner
and 2 error case of new owner
the latter 5 error can happen for act.OwnerAddress() == nil and act.OwnerAddress() != nil
so a total of 14 cases
can you add the remaining? (now it has 9)

Reply:
also can you group the test case in same order of code? that would be easier to look